### PR TITLE
fix: adjust BranchOp handling

### PR DESCRIPTION
**Pull request by gabriele-0201**
The changeset not only includes fixes for how uncompressed separators
are represented through BranchOp, but also contains a refactor
to better encapsulate the update logic of BranchOp.

Note: consume_and_update_until reports the following doc:
```
// If `stop_prefix_compression` has to be called, then the target becomes BRANCH_MERGE_THRESHOLD
// to minimize the amount of uncompressed items inserted in the node.
```

This was never really discussed before, but while working on the tests, I noticed how this was expected by the test `shared_prefix_collapse`. I thought, besides making sense leading to fewer edge cases, to explicitly state the behavior and ensure it effectively happens. It will be tested by multiple tests in follow-up PRs.

### CODE REVIEW

**Comment by gabriele-0201** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/615?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#616** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/616?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#615** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/615?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/615?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#614** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/614?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by rphmeier — APPROVED** [review]

File: nomt/src/beatree/ops/update/branch_updater.rs [hunk 1]
```diff
@@ -407,79 +357,40 @@ impl BranchUpdater {
...
-                            &mut self.ops,
-                            split_point,
-                        );
+        let (split_point, left_gauge) = match self.consume_and_update_until(0, midpoint) {
+            Ok((split_point, left_gauge)) => (split_point, left_gauge),
+            // If the current ops cannot reach the target and there is a cutoff
+            // return NeedsMerge with the relative cutoff.
+            Err(new_gauge) if self.cutoff.is_some() => {
+                self.gauge = new_gauge;
+                return DigestResult::NeedsMerge(self.cutoff.unwrap());
```
Comment (rphmeier) [inline nomt/src/beatree/ops/update/branch_updater.rs:367]: UNWRAP!

File: nomt/src/beatree/ops/update/branch_updater.rs [hunk 2]
```diff
@@ -510,6 +421,98 @@ impl BranchUpdater {
...
+                            // change the target requirement to minumize the number of non
+                            // compressed separators saved into one node
+                            target = BRANCH_MERGE_THRESHOLD;
+                        } else {
+                            break;
+                        }
+                    }
+                }
+                BranchOp::Update(update_pos, _) => {
+                    let key = self.base.as_ref().unwrap().key(update_pos);
```
Comment (rphmeier) [inline nomt/src/beatree/ops/update/branch_updater.rs:459]: UNWRAP!

File: nomt/src/beatree/ops/update/branch_updater.rs [hunk 3]
```diff
@@ -510,6 +421,98 @@ impl BranchUpdater {
...
+                            break;
+                        }
+                    }
+                }
+                BranchOp::KeepChunk(..) => {
+                    // Try to split the chunk to make it fit into the available space.
+                    // `try_split_keep_chunk` works on the gauge thus it accounts for a possible
+                    // stop of the prefix compression even if working on a KeepChunk operation
+                    let left_n_items = try_split_keep_chunk(
+                        self.base.as_ref().unwrap(),
```
Comment (rphmeier) [inline nomt/src/beatree/ops/update/branch_updater.rs:477]: UNWRAP!

File: nomt/src/beatree/ops/update/branch_updater.rs [hunk 4]
```diff
@@ -510,6 +421,98 @@ impl BranchUpdater {
...
+                        target,
+                        BRANCH_NODE_BODY_SIZE,
+                    );
+
+                    if left_n_items == 0 {
+                        // If no item from the chunk is capable of fitting,
+                        // then extract the first element from the chunk and repeat the loop
+                        // to see if `stop_prefix_compression` is activated
+                        extract_insert_from_keep_chunk(
+                            self.base.as_ref().unwrap(),
```
Comment (rphmeier) [inline nomt/src/beatree/ops/update/branch_updater.rs:490]: UNWRAP!

File: nomt/src/beatree/ops/update/branch_updater.rs (hunk 1 above)
Comment (gabriele-0201) [reply nomt/src/beatree/ops/update/branch_updater.rs:367]: Wow, yes, I totally forgot to comment all unwraps :sweat_smile:
